rig/routing_table/minimise.py

```diff
@@ -0,0 +1,557 @@
+However, packets with the key ``1111`` would be sent to cores 1 and 2 as only
+the higher priority entry has effect.
+
+Merging routing table entries
+=============================
+
+Routing tables can be minimised by merging together entries with equivalent
+routes. This is done by creating a new key-mask pair with an ``X`` wherever the
+key-mask pairs of any of the original entries differed.
+
+For example, merging of the entries::
+
+    0000 -> N
+    0001 -> N
+
+Would lead to the new entry:
+
+    000X -> N
+
+Which would match any of the keys matched by the original entries but no more.
+In contrast the merge of ``0001`` and ``0010`` would generate the new entry
+``00XX`` which would match keys matched by either of the original entries but
+also ``0000`` and ``0011``.
+
+Clearly, if we are to attempt to minimise tables such as::
+
+    0001 -> N
+    0010 -> N
+    0000 -> S, SE
+    0011 -> SE
+
+We need a set of rules for:
+
+ 1. Where merged entries are to be inserted into the table
+ 2. Which merges are allowed
+
+"Ordered Covering"
+==================
+
+The algorithm implemented here, "Ordered Covering", provides the following
+rules:
+
+ 1. Routing tables are to be kept sorted in increasing order of "generality",
+    that is the number of ``X``s in the entry. An entry with the key-mask pair
+    ``00XX`` must be placed below any entries with fewer ``X``s in their
+    key-mask pairs (e.g., below ``0000`` and ``000X``).
+
+    a) New entries must also be inserted below any entries of the same
+       generality. If ``XX00`` were already present in the table the new entry
+       ``0XX0`` must be inserted below it.
+
+ 2. The only merges allowed are those which:
+
+    a) would not cause one of the entries in the merge to be "hidden" below
+       an entry of lesser generality than the merged entry but which matched
+       any of the same keys. For example, merging ``0010`` and ``0001`` would
+       not be allowed if the new entry would be placed below the existing
+       entry ``000X`` as this would "hide" ``0001``.
+    b) would not cause an entry "contained" within an entry of higher
+       generality to be hidden by the insertion of a new entry. For example, if
+       the entry ``XXXX`` had been formed by merging the entries ``0011`` and
+       ``1100`` then merging of the entries ``1101`` and ``1110`` would not be
+       allowed as it would cause the entry ``11XX`` to be inserted above
+       ``XXXX`` in the table and would hide ``1100``.
+
+Following these rules ensures that the minimised table will be functionally
+equivalent to the original table provided that the original table was invariant
+under reordering OR was provided in increasing order of generality.
+"""
+from collections import defaultdict, namedtuple
+from rig.routing_table import RoutingTableEntry
+from six import itervalues
+
+
+def minimise(routing_table, target_length=1023):
```

Comment:
I'd say that target_length should default to 1024: If an assumption about table availability is to be made it should be made by end users, not in Rig. Hopefully we'll come up with something to solve this problem...